### PR TITLE
Improve discovery

### PR DESCRIPTION
**Pull request by lenak25**

### CODE REVIEW

**Review by assafmo — DISMISSED** [review]

File: docs/ARCHITECTURE.md [hunk 1]
```diff
@@ -369,33 +369,6 @@ Such as Ethereum listener, JsonRpcAPI etc.
 
 Register with Enigma.sol with all the required steps including Enclave Report.
 
-## Persistent Peer Discovery
```
Comment (assafmo) [inline docs/ARCHITECTURE.md:372]: Why isn't this needed anymore?

File: src/cli/cli_app.js [hunk 2]
```diff
@@ -112,43 +112,27 @@ class CLI {
...
       },
-      'discover': () =>{
-        this._node.tryConsistentDiscovery();
-      },
-      'inCount': () =>{
-        const cons = this._node.getAllInboundHandshakes();
-        console.log('---> inbound connections <---');
-        console.log(cons.length);
+      'getConnectedPeers': () =>{
+        let peers = this._node.getConnectedPeers();
```
Comment (assafmo) [inline src/cli/cli_app.js]: Using `const` instead of `let` can catch a lot of bugs.  
(Not needed here tough)

File: src/common/utils.js [hunk 3]
```diff
@@ -239,6 +238,34 @@ module.exports.peerBankSeedtoPeerInfo = function(seed, callback) {
   }
 };
 
+/**
+ * A promisified version of readFile
+ * @param {String} path - file path
+ * @return {Promise}
+ * */
```
Comment (assafmo) [inline src/common/utils.js]: These docs doesn't make any sense :woman_shrugging:

File: src/policy/policy.js [hunk 4]
```diff
@@ -9,26 +9,12 @@ class Policy extends EventEmitter {
     super();
 
     this._isStrict = true;
-    if(isStrict === false){
+    if (isStrict === false){
```
Comment (assafmo) [inline src/policy/policy.js]: This is good to know:  
https://stackoverflow.com/a/359509/1551596  
(And even makes it more confusing :rofl:)

File: src/worker/EnigmaNode.js [hunk 5]
```diff
@@ -499,63 +509,25 @@ class EnigmaNode extends EventEmitter {
...
-            // see constats.js for HANDSHAKE_OUTBOUND/INBOUND actions.
-            this.emit('notify', pongMsg);
-            onHandshake(err, peerInfo, ping, pongMsg);
-            return pongMsg.toNetworkStream();
-          })
-      );
+    }
+    this.dial(peerInfo, (connectionErr, connection) => {
+      if (!connectionErr) {
+        this.notify(peerInfo);
```
Comment (assafmo) [inline src/worker/EnigmaNode.js:484]: return?

File: src/worker/EnigmaNode.js [hunk 6]
```diff
@@ -499,63 +509,25 @@ class EnigmaNode extends EventEmitter {
...
+      }
+      onConnection(connectionErr);
```
Comment (assafmo) [inline src/worker/EnigmaNode.js]: else? This always gets called.

**Review by assafmo — DISMISSED** [review]

**Comment by lenak25** [comment]
@assafmo, this PR is ready for review

File: src/common/utils.js (hunk 3 above)
Comment (assafmo) [reply src/common/utils.js]: ?

File: src/worker/EnigmaNode.js [hunk 7]
```diff
@@ -420,16 +422,29 @@ class EnigmaNode extends EventEmitter {
...
   }
+  /**
+   * Dial at some protocol and delegate the handling of that connection
+   * @param {PeerInfo} peerInfo ,  the peer we wish to dial to
+   * @param {Function} onConnection receives (err,connection) =>{}
+   */
+  dial(peerInfo, onConnection) {
+    if (peerInfo.id.toB58String() === this.getSelfIdB58Str()) {
+      this._logger.error('[-] Error : ' + MSG_STATUS.ERR_SELF_DIAL);
+    }
```
Comment (assafmo) [inline src/worker/EnigmaNode.js:411]: I think it's unnecessary.  
`onConnection` will get the error and can handle it from there.  
This way error handling for dial will always happen in only one place.

Screenshot from 2019-11-19 15-01-25

File: src/worker/controller/actions/connectivity/BootstrapDiscoveredAction.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+  execute(params) {
+    params = params.params;
+    const otherPeer = params.peer;
+
+    // Connect to a bootstrap only if there are no active connections
+    if (this._controller.engNode().arePeersConnected()) {
+      return;
+    }
+    this._controller.engNode().connectToBootstrap(otherPeer);
```
Comment (assafmo) [inline src/worker/controller/actions/connectivity/BootstrapDiscoveredAction.js]: Is this like this?:  
https://github.com/ipfs/js-ipfs/pull/1914/files

**Review by assafmo — DISMISSED** [review]

**Review by assafmo — APPROVED** [review]
:100:

**Pull request merged**